File: tests/remote/mockSheetsAdapter.test.ts
import { describe, it, expect } from "vitest";
import { MockSheetsAdapter } from "@/remote/mock/mockSheetsAdapter";
import { newAssignment, newEvent } from "@/domain/models";
import {
  AssignmentStatus,
  EventType,
  NotificationCategory,
  RideLeg,
  VisibilityScope,
} from "@/domain/enums";
import { defaultGroupSharedConfig } from "@/domain/config";
import type { NotificationEntry } from "@/domain/models";

describe("MockSheetsAdapter", () => {
  it("events write then read preserves every field", async () => {
    const a = new MockSheetsAdapter();
    const sheetId = await a.ensureSheetExists("g1", "Group One");
    const event = newEvent({
      eventId: "e-roundtrip",
      groupId: "g1",
      childId: "c1",
      title: "Swimming",
      eventType: EventType.SPORTS,
      locationName: "Pool",
      startDateTime: 1_714_500_000_000,
      endDateTime: 1_714_503_600_000,
      needsRide: true,
      createdByParentId: "p1",
      visibilityScope: VisibilityScope.GROUP,
      updatedAt: 111,
    });
    await a.writeEvents(sheetId, [event]);
    const read = await a.readEvents(sheetId);
    expect(read).toHaveLength(1);
    expect(read[0]).toEqual(event);
  });

  it("assignments write then read preserves every field", async () => {
    const a = new MockSheetsAdapter();
    const sheetId = await a.ensureSheetExists("g1", "Group One");
    const assign = newAssignment({
      assignmentId: "a-1",
      eventId: "e-1",
      driverParentId: "p2",
      rideLeg: RideLeg.FROM,
      assignmentStatus: AssignmentStatus.CONFIRMED,
      notes: "carpool with neighbor",
      claimedAt: 1_714_500_000_000,
      completedAt: null,
      updatedAt: 2,
    });
    await a.writeAssignments(sheetId, [assign]);
    const read = await a.readAssignments(sheetId);
    expect(read).toHaveLength(1);
    expect(read[0]).toEqual(assign);
  });

  it("groupConfig write then read preserves every field", async () => {
    const a = new MockSheetsAdapter();
    const sheetId = await a.ensureSheetExists("g1", "Group One");
    const cfg = {
      ...defaultGroupSharedConfig,
      groupName: "Renamed",
      timezone: "Europe/London",
      enableRideSharing: false,
      maxFutureEventMonths: 3,
    };
    await a.writeGroupConfig(sheetId, cfg);
    expect(await a.readGroupConfig(sheetId)).toEqual(cfg);
  });

  it("notifications append accumulates entries", async () => {
    const a = new MockSheetsAdapter();
    const sheetId = await a.ensureSheetExists("g1", "Group One");
    const n1: NotificationEntry = {
      notificationId: "n1", groupId: "g1", eventId: "e1", assignmentId: null,
      triggeredByParentId: "p1", message: "Ride claimed",
      category: NotificationCategory.RIDE_CLAIMED, createdAt: 100,
    };
    const n2 = { ...n1, notificationId: "n2", message: "Ride completed",
      category: NotificationCategory.RIDE_COMPLETED, createdAt: 200 };
    await a.appendNotification(sheetId, n1);
    await a.appendNotification(sheetId, n2);
    const read = await a.readNotifications(sheetId);
    expect(read).toEqual(expect.arrayContaining([n1, n2]));
  });

  it("ensureSheetExists is idempotent per groupId", async () => {
    const a = new MockSheetsAdapter();
    const first = await a.ensureSheetExists("g1", "Group One");
    const second = await a.ensureSheetExists("g1", "Group One renamed");
    expect(second).toBe(first);
  });

  it("new sheet seeds with Piano and Soccer events", async () => {
    const a = new MockSheetsAdapter();
    const sheetId = "seed-sheet-1";
    const events = await a.readEvents(sheetId);
    expect(events.length).toBeGreaterThan(0);
    expect(events.some((e) => e.title === "Piano Lesson")).toBe(true);
  });
});

File: app/src/test/java/i/drive/kids/data/remote/mock/MockSheetsAdapterTest.kt
package i.drive.kids.data.remote.mock

import i.drive.kids.config.GroupSharedConfig
import i.drive.kids.domain.model.AssignmentStatus
import i.drive.kids.domain.model.Event
import i.drive.kids.domain.model.EventType
import i.drive.kids.domain.model.NotificationCategory
import i.drive.kids.domain.model.NotificationEntry
import i.drive.kids.domain.model.RideAssignment
import i.drive.kids.domain.model.RideLeg
import i.drive.kids.domain.model.VisibilityScope
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * SheetRowMapperTest equivalent — exercises all Sheet tabs (Events, Assignments,
 * Parents, GroupConfig, Notifications) through the mock adapter to confirm
 * read/write is a lossless identity on the in-memory representation.
 */
class MockSheetsAdapterTest {

    private fun newAdapter() = MockSheetsAdapter()

    @Test
    fun `events write then read preserves every field`() = runBlocking {
        val adapter = newAdapter()
        val sheetId = adapter.ensureSheetExists("g1", "Group One")
        val event = Event(
            eventId = "e-roundtrip",
            groupId = "g1",
            childId = "c1",
            title = "Swimming",
            eventType = EventType.SPORTS,
            locationName = "Pool",
            startDateTime = 1_714_500_000_000L,
            endDateTime = 1_714_503_600_000L,
            needsRide = true,
            createdByParentId = "p1",
            visibilityScope = VisibilityScope.GROUP,
            updatedAt = 111L
        )
        adapter.writeEvents(sheetId, listOf(event))

        val read = adapter.readEvents(sheetId)

        assertEquals(1, read.size)
        assertEquals(event, read[0])
    }

    @Test
    fun `assignments write then read preserves every field`() = runBlocking {
        val adapter = newAdapter()
        val sheetId = adapter.ensureSheetExists("g1", "Group One")
        val a = RideAssignment(
            assignmentId = "a-1",
            eventId = "e-1",
            driverParentId = "p2",
            rideLeg = RideLeg.FROM,
            assignmentStatus = AssignmentStatus.CONFIRMED,
            notes = "carpool with neighbor",
            claimedAt = 1_714_500_000_000L,
            completedAt = null,
            updatedAt = 2L
        )
        adapter.writeAssignments(sheetId, listOf(a))

        val read = adapter.readAssignments(sheetId)

        assertEquals(1, read.size)
        assertEquals(a, read[0])
    }

    @Test
    fun `groupConfig write then read preserves every field`() = runBlocking {
        val adapter = newAdapter()
        val sheetId = adapter.ensureSheetExists("g1", "Group One")
        val cfg = GroupSharedConfig(
            groupName = "Renamed",
            timezone = "Europe/London",
            enableRideSharing = false,
            maxFutureEventMonths = 3
        )
        adapter.writeGroupConfig(sheetId, cfg)

        val read = adapter.readGroupConfig(sheetId)

        assertEquals(cfg, read)
    }

    @Test
    fun `notifications append accumulates entries`() = runBlocking {
        val adapter = newAdapter()
        val sheetId = adapter.ensureSheetExists("g1", "Group One")
        val n1 = NotificationEntry(
            notificationId = "n1", groupId = "g1", eventId = "e1", assignmentId = null,
            triggeredByParentId = "p1", message = "Ride claimed",
            category = NotificationCategory.RIDE_CLAIMED, createdAt = 100L
        )
        val n2 = n1.copy(notificationId = "n2", message = "Ride completed",
            category = NotificationCategory.RIDE_COMPLETED, createdAt = 200L)

        adapter.appendNotification(sheetId, n1)
        adapter.appendNotification(sheetId, n2)

        val read = adapter.readNotifications(sheetId)

        assertTrue(read.containsAll(listOf(n1, n2)))
    }

    @Test
    fun `ensureSheetExists is idempotent per groupId`() = runBlocking {
        val adapter = newAdapter()
        val first = adapter.ensureSheetExists("g1", "Group One")
        val second = adapter.ensureSheetExists("g1", "Group One renamed")
        assertEquals(first, second)
        assertNotNull(first)
    }

    @Test
    fun `new sheet seeds with piano lesson and soccer events`() = runBlocking {
        val adapter = newAdapter()
        val sheetId = "seed-sheet-1"
        val events = adapter.readEvents(sheetId)
        assertTrue("should seed at least one event", events.isNotEmpty())
        assertTrue(events.any { it.title == "Piano Lesson" })
    }
}
